### PR TITLE
[SPARK-23195] [SQL] Keep the Hint of Cached Data

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
The broadcast hint of the cached plan is lost if we cache the plan. This PR is to correct it.

```Scala
  val df1 = spark.createDataFrame(Seq((1, "4"), (2, "2"))).toDF("key", "value")
  val df2 = spark.createDataFrame(Seq((1, "1"), (2, "2"))).toDF("key", "value")
  broadcast(df2).cache()
  df2.collect()
  val df3 = df1.join(df2, Seq("key"), "inner")
```

## How was this patch tested?
Added a test.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala [hunk 1]
```diff
@@ -110,15 +110,39 @@ class BroadcastJoinSuite extends QueryTest with SQLTestUtils {
...
+        val df2 = spark.createDataFrame(Seq((1, "1"), (2, "2"))).toDF("key", "value")
+        df2.cache()
+        val df3 = df1.join(broadcast(df2), Seq("key"), "inner")
+        val numBroadCastHashJoin = df3.queryExecution.executedPlan.collect {
+          case b: BroadcastHashJoinExec => b
+        }.size
+        assert(numBroadCastHashJoin === 1)
+      }
+    } finally {
+      spark.catalog.clearCache()
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala]: We have to clear the cache. We have another bug to fix in the cache. Will submit another PR to fix the cache plan matching.

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala]: If we have to clear the cache, can we add `clearCache` into `afterEach` in general instead of adding this case-by-case?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala]: This test suite is not for testing cache. Thus, it is fine to do it for these two test cases.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by gatorsmile** [comment]
cc @cloud-fan @sameeragarwal @jiangxb1987

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala]: do you mean we can remove the cache cleaning here after you fix that bug?

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/joins/BroadcastJoinSuite.scala]: Yeah. That should be a separate bug.
